From 8991c655fb8da6052d8008970fc2639c98430186 Mon Sep 17 00:00:00 2001
From: Mike Drob <mdrob@apache.org>
Date: Thu, 20 Jul 2017 19:13:12 -0500
Subject: [PATCH] HBASE-18419 update ITIngestWithMOB for immutable
 columndescriptors

---
 .../hadoop/hbase/IntegrationTestIngestWithMOB.java | 29 ++++++++++++----------
 1 file changed, 16 insertions(+), 13 deletions(-)

diff --git a/hbase-it/src/test/java/org/apache/hadoop/hbase/IntegrationTestIngestWithMOB.java b/hbase-it/src/test/java/org/apache/hadoop/hbase/IntegrationTestIngestWithMOB.java
index cd9e355b20..5bbb12b14b 100644
--- a/hbase-it/src/test/java/org/apache/hadoop/hbase/IntegrationTestIngestWithMOB.java
+++ b/hbase-it/src/test/java/org/apache/hadoop/hbase/IntegrationTestIngestWithMOB.java
@@ -25,6 +25,9 @@ import java.util.List;
 import org.apache.commons.cli.CommandLine;
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.hbase.client.Admin;
+import org.apache.hadoop.hbase.client.ColumnFamilyDescriptor;
+import org.apache.hadoop.hbase.client.ColumnFamilyDescriptorBuilder;
+import org.apache.hadoop.hbase.client.Connection;
 import org.apache.hadoop.hbase.client.ConnectionFactory;
 import org.apache.hadoop.hbase.testclassification.IntegrationTests;
 import org.apache.hadoop.hbase.util.Bytes;
@@ -113,20 +116,20 @@ public class IntegrationTestIngestWithMOB extends IntegrationTestIngest {
     super.initTable();
 
     TableName tableName = getTablename();
-    Admin admin = ConnectionFactory.createConnection().getAdmin();
-    HTableDescriptor tableDesc = admin.getTableDescriptor(tableName);
-    LOG.info("Disabling table " + getTablename());
-    admin.disableTable(tableName);
-    for (HColumnDescriptor columnDescriptor : tableDesc.getFamilies()) {
-      if(Arrays.equals(columnDescriptor.getName(), mobColumnFamily)) {
-        columnDescriptor.setMobEnabled(true);
-        columnDescriptor.setMobThreshold((long) threshold);
-        admin.modifyColumnFamily(tableName, columnDescriptor);
-      }
+    try (Connection connection = ConnectionFactory.createConnection();
+         Admin admin = connection.getAdmin()) {
+      HTableDescriptor tableDesc = admin.getTableDescriptor(tableName);
+      LOG.info("Disabling table " + getTablename());
+      admin.disableTable(tableName);
+      ColumnFamilyDescriptor mobColumn = tableDesc.getColumnFamily(mobColumnFamily);
+      ColumnFamilyDescriptor cfd = ColumnFamilyDescriptorBuilder.newBuilder(mobColumn)
+        .setMobEnabled(true)
+        .setMobThreshold((long) threshold)
+        .build();
+      admin.modifyColumnFamily(tableName, cfd);
+      LOG.info("Enabling table " + getTablename());
+      admin.enableTable(tableName);
     }
-    LOG.info("Enabling table " + getTablename());
-    admin.enableTable(tableName);
-    admin.close();
   }
 
   @Override
-- 
2.13.3

