diff --git src/main/java/org/apache/hadoop/hbase/HConstants.java src/main/java/org/apache/hadoop/hbase/HConstants.java
index 5120a3c..4045c7f 100644
--- src/main/java/org/apache/hadoop/hbase/HConstants.java
+++ src/main/java/org/apache/hadoop/hbase/HConstants.java
@@ -69,6 +69,9 @@ public final class HConstants {
   /** Cluster is in distributed mode or not */
   public static final String CLUSTER_DISTRIBUTED = "hbase.cluster.distributed";
 
+  /** Config for pluggable load balancers */
+  public static final String HBASE_MASTER_LOADBALANCER_CLASS = "hbase.master.loadbalancer.class";
+
   /** Cluster is standalone or pseudo-distributed */
   public static final String CLUSTER_IS_LOCAL = "false";
 
diff --git src/main/java/org/apache/hadoop/hbase/master/LoadBalancerFactory.java src/main/java/org/apache/hadoop/hbase/master/LoadBalancerFactory.java
index 89685bb..3a22e46 100644
--- src/main/java/org/apache/hadoop/hbase/master/LoadBalancerFactory.java
+++ src/main/java/org/apache/hadoop/hbase/master/LoadBalancerFactory.java
@@ -21,6 +21,7 @@
 package org.apache.hadoop.hbase.master;
 
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HConstants;
 import org.apache.hadoop.util.ReflectionUtils;
 
 /**
@@ -36,7 +37,9 @@ public class LoadBalancerFactory {
   public static LoadBalancer getLoadBalancer(Configuration conf) {
 
     // Create the balancer
-    Class<? extends LoadBalancer> balancerKlass = conf.getClass("hbase.maser.loadBalancer.class",DefaultLoadBalancer.class, LoadBalancer.class);
+    Class<? extends LoadBalancer> balancerKlass = conf.getClass(
+        HConstants.HBASE_MASTER_LOADBALANCER_CLASS,
+        DefaultLoadBalancer.class, LoadBalancer.class);
     return ReflectionUtils.newInstance(balancerKlass, conf);
 
   }
