Index: src/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java
===================================================================
--- src/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java	(revision 705658)
+++ src/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java	(working copy)
@@ -295,6 +295,7 @@
       new Sleeper(this.msgInterval, this.stopRequested);
     try {
       init(reportForDuty(sleeper));
+      getRootRegion(sleeper);
       long lastMsg = 0;
       // Now ask master what it wants us to do and tell it what we have done
       for (int tries = 0; !stopRequested.get() && isHealthy();) {
@@ -535,6 +536,22 @@
       throw ex;
     }
   }
+  
+  private void getRootRegion(Sleeper sleeper) {
+    // Try to get the root region location from the master. 
+    while (!haveRootRegion.get()) {
+      HServerAddress rootServer = hbaseMaster.getRootRegionLocation();
+      if (rootServer == null) {
+        sleeper.sleep();
+      } else {
+        // By setting the root region location, we bypass the wait imposed on
+        // HTable for all regions being assigned.
+        this.connection.setRootRegionLocation(
+            new HRegionLocation(HRegionInfo.ROOT_REGIONINFO, rootServer));
+        haveRootRegion.set(true);
+      }
+    }
+  }
 
   /**
    * Report the status of the server. A server is online once all the startup 
@@ -633,17 +650,6 @@
    * the end of the main HRegionServer run loop.
    */
   private void housekeeping() {
-    // Try to get the root region location from the master. 
-    if (!haveRootRegion.get()) {
-      HServerAddress rootServer = hbaseMaster.getRootRegionLocation();
-      if (rootServer != null) {
-        // By setting the root region location, we bypass the wait imposed on
-        // HTable for all regions being assigned.
-        this.connection.setRootRegionLocation(
-            new HRegionLocation(HRegionInfo.ROOT_REGIONINFO, rootServer));
-        haveRootRegion.set(true);
-      }
-    }
     // If the todo list has > 0 messages, iterate looking for open region
     // messages. Send the master a message that we're working on its
     // processing so it doesn't assign the region elsewhere.
