diff --git hbase-client/src/main/java/org/apache/hadoop/hbase/client/ServerCallable.java hbase-client/src/main/java/org/apache/hadoop/hbase/client/ServerCallable.java
index fe25272..994efa6 100644
--- hbase-client/src/main/java/org/apache/hadoop/hbase/client/ServerCallable.java
+++ hbase-client/src/main/java/org/apache/hadoop/hbase/client/ServerCallable.java
@@ -200,8 +200,8 @@ public abstract class ServerCallable<T> implements Callable<T> {
         // If the server is dead, we need to wait a little before retrying, to give
         //  a chance to the regions to be
         expectedSleep = ConnectionUtils.getPauseTime(pause, tries);
-        if (expectedSleep < MIN_WAIT_DEAD_SERVER &&
-            getConnection().isDeadServer(location.getServerName())){
+        if (expectedSleep < MIN_WAIT_DEAD_SERVER 
+            && (location == null || getConnection().isDeadServer(location.getServerName()))) {
           expectedSleep = ConnectionUtils.addJitter(MIN_WAIT_DEAD_SERVER, 0.10f);
         }
 
