diff --git hbase-client/src/main/java/org/apache/hadoop/hbase/zookeeper/ZooKeeperWatcher.java hbase-client/src/main/java/org/apache/hadoop/hbase/zookeeper/ZooKeeperWatcher.java
index 6cdd6da..cf6e1df 100644
--- hbase-client/src/main/java/org/apache/hadoop/hbase/zookeeper/ZooKeeperWatcher.java
+++ hbase-client/src/main/java/org/apache/hadoop/hbase/zookeeper/ZooKeeperWatcher.java
@@ -111,6 +111,9 @@ public class ZooKeeperWatcher implements Watcher, Abortable, Closeable {
   public String recoveringRegionsZNode;
   // znode containing namespace descriptors
   public static String namespaceZNode = "namespace";
+  // znode for tracking upgrade (to 0.96). It is created while doing upgrade, and is deleted once
+  // master has processed all the old logs when starting up.
+  public static String upgradeInprocessZnode;
 
 
   // Certain ZooKeeper nodes need to be world-readable
@@ -237,6 +240,8 @@ public class ZooKeeperWatcher implements Watcher, Abortable, Closeable {
         conf.get("zookeeper.znode.recovering.regions", "recovering-regions"));
     namespaceZNode = ZKUtil.joinZNode(baseZNode,
         conf.get("zookeeper.znode.namespace", "namespace"));
+    upgradeInprocessZnode = ZKUtil.joinZNode(baseZNode,
+        conf.get("zookeeper.znode.upgradeInprogress", "upgradeInprogress"));
   }
 
   /**
diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java
index e8c554e..4b2fa06 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java
@@ -823,6 +823,13 @@ MasterServices, Server {
     // handle other failed servers in SSH in order to start up master node ASAP
     Set<ServerName> previouslyFailedServers = this.fileSystemManager
         .getFailedServersFromLogFolders();
+    // if doing upgrade, ensure we split all logs before moving to assigning regions.
+    // Ideally, this should happen only once.
+    if (ZKUtil.checkExists(zooKeeper, ZooKeeperWatcher.upgradeInprocessZnode) != -1) {
+      this.fileSystemManager.splitLog(previouslyFailedServers);
+      // once done, delete the upgradeInprogressZnode
+      ZKUtil.deleteNode(zooKeeper, ZooKeeperWatcher.upgradeInprocessZnode);
+    }
 
     // remove stale recovering regions from previous run
     this.fileSystemManager.removeStaleRecoveringRegionsFromZK(previouslyFailedServers);
diff --git hbase-server/src/main/java/org/apache/hadoop/hbase/util/ZKDataMigrator.java hbase-server/src/main/java/org/apache/hadoop/hbase/util/ZKDataMigrator.java
index 311a4ed..ab90533 100644
--- hbase-server/src/main/java/org/apache/hadoop/hbase/util/ZKDataMigrator.java
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/util/ZKDataMigrator.java
@@ -45,11 +45,11 @@ import org.apache.zookeeper.KeeperException.NoNodeException;
 public class ZKDataMigrator extends Configured implements Tool {
 
   private static final Log LOG = LogFactory.getLog(ZKDataMigrator.class);
+  ZooKeeperWatcher zkw = null;
 
   @Override
   public int run(String[] as) throws Exception {
     Configuration conf = getConf();
-    ZooKeeperWatcher zkw = null;
     try {
       zkw = new ZooKeeperWatcher(getConf(), "Migrate ZK data to PB.",
         new ZKDataMigratorAbortable());
@@ -120,6 +120,7 @@ public class ZKDataMigrator extends Configured implements Tool {
           ZKUtil.deleteNodeRecursively(zkw, childPath);
         }
       }
+      createUpgradeInprogressZnode();
     } catch (Exception e) {
       LOG.error("Got exception while updating znodes ", e);
       throw new IOException(e);
@@ -131,6 +132,10 @@ public class ZKDataMigrator extends Configured implements Tool {
     return 0;
   }
 
+  private void createUpgradeInprogressZnode() throws KeeperException {
+    ZKUtil.createAndFailSilent(zkw, ZooKeeperWatcher.upgradeInprocessZnode);
+  }
+
   private void checkAndMigrateTableStatesToPB(ZooKeeperWatcher zkw) throws KeeperException {
     List<String> tables = ZKUtil.listChildrenNoWatch(zkw, zkw.tableZNode);
     if (tables == null) {
