diff --git a/hbase-protocol/src/main/protobuf/ZooKeeper.proto b/hbase-protocol/src/main/protobuf/ZooKeeper.proto
index 69a5c01..e20e17b 100644
--- a/hbase-protocol/src/main/protobuf/ZooKeeper.proto
+++ b/hbase-protocol/src/main/protobuf/ZooKeeper.proto
@@ -44,7 +44,7 @@ message Master {
 }
 
 /**
- * Content of the '/hbase/shutdown', cluster state, znode.
+ * Content of the '/hbase/running', cluster state, znode.
  */
 message ClusterUp {
   // If this znode is present, cluster is up.  Currently
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/zookeeper/ZooKeeperWatcher.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/zookeeper/ZooKeeperWatcher.java
index 4adda83..48fe695 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/zookeeper/ZooKeeperWatcher.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/zookeeper/ZooKeeperWatcher.java
@@ -204,7 +204,7 @@ public class ZooKeeperWatcher implements Watcher, Abortable, Closeable {
     backupMasterAddressesZNode = ZKUtil.joinZNode(baseZNode,
         conf.get("zookeeper.znode.backup.masters", "backup-masters"));
     clusterStateZNode = ZKUtil.joinZNode(baseZNode,
-        conf.get("zookeeper.znode.state", "shutdown"));
+        conf.get("zookeeper.znode.state", "running"));
     assignmentZNode = ZKUtil.joinZNode(baseZNode,
         conf.get("zookeeper.znode.unassigned", "unassigned"));
     tableZNode = ZKUtil.joinZNode(baseZNode,
