From bc0e9ce8f462d8b3f4ca67ad80d455edb39605bf Mon Sep 17 00:00:00 2001
From: Michael Stack <stack@apache.org>
Date: Wed, 31 Jan 2018 17:32:39 -0800
Subject: [PATCH] HBASE-19908 TestCoprocessorShortCircuitRPC Timeout....

---
 .../hadoop/hbase/coprocessor/TestCoprocessorShortCircuitRPC.java      | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestCoprocessorShortCircuitRPC.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestCoprocessorShortCircuitRPC.java
index c3f7e80ba9..6cff379a44 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestCoprocessorShortCircuitRPC.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestCoprocessorShortCircuitRPC.java
@@ -28,7 +28,7 @@ import org.apache.hadoop.hbase.SharedConnection;
 import org.apache.hadoop.hbase.client.Connection;
 import org.apache.hadoop.hbase.client.ConnectionUtils;
 import org.apache.hadoop.hbase.testclassification.CoprocessorTests;
-import org.apache.hadoop.hbase.testclassification.SmallTests;
+import org.apache.hadoop.hbase.testclassification.MediumTests;
 import org.junit.AfterClass;
 import org.junit.BeforeClass;
 import org.junit.ClassRule;
@@ -41,7 +41,7 @@ import org.junit.rules.TestName;
  * Ensure Coprocessors get ShortCircuit Connections when they get a Connection from their
  * CoprocessorEnvironment.
  */
-@Category({CoprocessorTests.class, SmallTests.class})
+@Category({CoprocessorTests.class, MediumTests.class})
 public class TestCoprocessorShortCircuitRPC {
 
   @ClassRule
-- 
2.11.0 (Apple Git-81)
