Index: hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/HBaseServer.java
===================================================================
--- hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/HBaseServer.java	(revision 1399501)
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/HBaseServer.java	(working copy)
@@ -2145,7 +2145,7 @@
    * done in chunks of this size. Most RPC requests and responses would be
    * be smaller.
    */
-  private static int NIO_BUFFER_LIMIT = 8*1024; //should not be more than 64KB.
+  private static int NIO_BUFFER_LIMIT = 64 * 1024; //should not be more than 64KB.
 
   /**
    * This is a wrapper around {@link java.nio.channels.WritableByteChannel#write(java.nio.ByteBuffer)}.
