diff --git a/src/main/java/org/apache/hadoop/hbase/master/HMaster.java b/src/main/java/org/apache/hadoop/hbase/master/HMaster.java
index 26fe694..449fd9e 100644
--- a/src/main/java/org/apache/hadoop/hbase/master/HMaster.java
+++ b/src/main/java/org/apache/hadoop/hbase/master/HMaster.java
@@ -1437,6 +1437,7 @@ implements HMasterInterface, HMasterRegionInterface, MasterServices, Server {
    * @see org.apache.hadoop.hbase.master.HMasterCommandLine
    */
   public static void main(String [] args) throws Exception {
+	VersionInfo.logVersion();
     new HMasterCommandLine(HMaster.class).doMain(args);
   }
 }
diff --git a/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java b/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java
index 89eae4b..55f1bb0 100644
--- a/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java
+++ b/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java
@@ -125,6 +125,7 @@ import org.apache.hadoop.hbase.util.InfoServer;
 import org.apache.hadoop.hbase.util.Pair;
 import org.apache.hadoop.hbase.util.Sleeper;
 import org.apache.hadoop.hbase.util.Threads;
+import org.apache.hadoop.hbase.util.VersionInfo;
 import org.apache.hadoop.hbase.zookeeper.ClusterStatusTracker;
 import org.apache.hadoop.hbase.zookeeper.ZKUtil;
 import org.apache.hadoop.hbase.zookeeper.ZooKeeperNodeTracker;
@@ -2949,6 +2950,7 @@ public class HRegionServer implements HRegionInterface, HBaseRPCErrorHandler,
    * @see org.apache.hadoop.hbase.regionserver.HRegionServerCommandLine
    */
   public static void main(String[] args) throws Exception {
+	VersionInfo.logVersion();
     Configuration conf = HBaseConfiguration.create();
     @SuppressWarnings("unchecked")
     Class<? extends HRegionServer> regionServerClass = (Class<? extends HRegionServer>) conf
diff --git a/src/main/java/org/apache/hadoop/hbase/rest/Main.java b/src/main/java/org/apache/hadoop/hbase/rest/Main.java
index 722b061..55b3562 100644
--- a/src/main/java/org/apache/hadoop/hbase/rest/Main.java
+++ b/src/main/java/org/apache/hadoop/hbase/rest/Main.java
@@ -31,6 +31,7 @@ import org.apache.commons.logging.LogFactory;
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.hbase.HBaseConfiguration;
 import org.apache.hadoop.hbase.rest.filter.GzipFilter;
+import org.apache.hadoop.hbase.util.VersionInfo;
 
 import java.util.List;
 import java.util.ArrayList;
@@ -68,6 +69,7 @@ public class Main implements Constants {
   public static void main(String[] args) throws Exception {
     Log LOG = LogFactory.getLog("RESTServer");
 
+	VersionInfo.logVersion();
     Configuration conf = HBaseConfiguration.create();
     RESTServlet servlet = RESTServlet.getInstance(conf);
 
diff --git a/src/main/java/org/apache/hadoop/hbase/thrift/ThriftServer.java b/src/main/java/org/apache/hadoop/hbase/thrift/ThriftServer.java
index be5f36d..92b6d00 100644
--- a/src/main/java/org/apache/hadoop/hbase/thrift/ThriftServer.java
+++ b/src/main/java/org/apache/hadoop/hbase/thrift/ThriftServer.java
@@ -70,6 +70,7 @@ import org.apache.hadoop.hbase.thrift.generated.TRegionInfo;
 import org.apache.hadoop.hbase.thrift.generated.TRowResult;
 import org.apache.hadoop.hbase.thrift.generated.TScan;
 import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.hadoop.hbase.util.VersionInfo;
 import org.apache.thrift.TException;
 import org.apache.thrift.protocol.TBinaryProtocol;
 import org.apache.thrift.protocol.TCompactProtocol;
@@ -1022,7 +1023,6 @@ public class ThriftServer {
       serverArgs.processor(processor);
       serverArgs.protocolFactory(protocolFactory);
       serverArgs.transportFactory(transportFactory);
-
       LOG.info("starting HBase ThreadPool Thrift server on " + listenAddress + ":" + Integer.toString(listenPort));
       server = new TThreadPoolServer(serverArgs);
     }
@@ -1035,6 +1035,7 @@ public class ThriftServer {
    * @throws Exception
    */
   public static void main(String [] args) throws Exception {
+	VersionInfo.logVersion();
     doMain(args);
   }
 }
diff --git a/src/main/java/org/apache/hadoop/hbase/util/VersionInfo.java b/src/main/java/org/apache/hadoop/hbase/util/VersionInfo.java
index 24e98df..e13f197 100644
--- a/src/main/java/org/apache/hadoop/hbase/util/VersionInfo.java
+++ b/src/main/java/org/apache/hadoop/hbase/util/VersionInfo.java
@@ -20,13 +20,17 @@
 
 package org.apache.hadoop.hbase.util;
 
+import org.apache.commons.logging.LogFactory;
 import org.apache.hadoop.hbase.VersionAnnotation;
+import org.apache.hadoop.hbase.master.HMaster;
+import org.apache.commons.logging.Log;
 
 /**
  * This class finds the package info for hbase and the VersionAnnotation
  * information.  Taken from hadoop.  Only name of annotation is different.
  */
 public class VersionInfo {
+  private static final Log LOG = LogFactory.getLog(VersionInfo.class.getName());
   private static Package myPackage;
   private static VersionAnnotation version;
 
@@ -82,6 +86,12 @@ public class VersionInfo {
   public static String getUrl() {
     return version != null ? version.url() : "Unknown";
   }
+  
+  public static void logVersion(){
+	  LOG.info("HBase " + getVersion());
+	  LOG.info("Subversion " + getUrl() + " -r " + getRevision());
+	  LOG.info("Compiled by " + getUser() + " on " + getDate());
+  }
 
   public static void main(String[] args) {
     System.out.println("HBase " + getVersion());
