Index: F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/RSDumpServlet.java
===================================================================
--- F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/RSDumpServlet.java	(revision 7615)
+++ F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/RSDumpServlet.java	(revision 7634)
@@ -80,6 +80,27 @@
     long tailKb = getTailKbParam(request);
     LogMonitoring.dumpTailOfLogs(out, tailKb);
     
+    out.println("\n\nRS Queue:");
+    out.println(LINE);
+    dumpQueue(hrs, out);
+    
     out.flush();
-  }  
+  }
+  
+  private void dumpQueue(HRegionServer hrs, PrintWriter out)
+      throws IOException {
+    if( hrs.isShowQueueDump() ) {
+      
+      // 1. Print out Compaction/Split Queue
+      out.println("Compaction/Split Queue summary: " 
+          + hrs.compactSplitThread.toString() );
+      out.println(hrs.compactSplitThread.dumpQueue());
+      
+      // 2. Print out flush Queue
+      out.println("\nFlush Queue summary: "
+          + hrs.cacheFlusher.toString());
+      out.println(hrs.cacheFlusher.dumpQueue());
+    }
+  }
+  
 }
Index: F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/MemStoreFlusher.java
===================================================================
--- F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/MemStoreFlusher.java	(revision 7615)
+++ F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/MemStoreFlusher.java	(revision 7634)
@@ -466,6 +466,26 @@
     }
   }
 
+  @Override
+  public String toString() {
+    return "flush_queue="
+        + flushQueue.size();
+  }
+  
+  public String dumpQueue() {
+    StringBuilder queueList = new StringBuilder();
+    queueList.append("Flush Queue Queue dump:\n");
+    queueList.append("  Flush Queue:\n");
+    java.util.Iterator<FlushQueueEntry> it = flushQueue.iterator();
+    
+    while(it.hasNext()){
+      queueList.append("    "+it.next().toString());
+      queueList.append("\n");
+    }
+    
+    return queueList.toString();
+  }
+  
   interface FlushQueueEntry extends Delayed {}
 
   /**
Index: F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/CompactSplitThread.java
===================================================================
--- F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/CompactSplitThread.java	(revision 7615)
+++ F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/CompactSplitThread.java	(revision 7634)
@@ -19,6 +19,8 @@
  */
 package org.apache.hadoop.hbase.regionserver;
 
+import java.util.Iterator;
+import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.Executors;
 import java.util.concurrent.PriorityBlockingQueue;
 import java.util.concurrent.RejectedExecutionException;
@@ -142,6 +144,40 @@
             : largeCompactions.getQueue().size())
         + ", split_queue=" + splits.getQueue().size();
   }
+  
+  public String dumpQueue() {
+    StringBuffer queueLists = new StringBuffer();
+    queueLists.append("Compaction/Split Queue dump:\n");
+    queueLists.append("  LargeCompation Queue:\n");
+    BlockingQueue<Runnable> lq = largeCompactions.getQueue();
+    Iterator it = lq.iterator();
+    while(it.hasNext()){
+      queueLists.append("    "+it.next().toString());
+      queueLists.append("\n");
+    }
+    
+    if( smallCompactions != null ){
+      queueLists.append("\n");
+      queueLists.append("  SmallCompation Queue:\n");
+      lq = smallCompactions.getQueue();
+      it = lq.iterator();
+      while(it.hasNext()){
+        queueLists.append("    "+it.next().toString());
+        queueLists.append("\n");
+      }
+    }
+    
+    queueLists.append("\n");
+    queueLists.append("  Split Queue:\n");
+    lq = splits.getQueue();
+    it = lq.iterator();
+    while(it.hasNext()){
+      queueLists.append("    "+it.next().toString());
+      queueLists.append("\n");
+    }
+    
+    return queueLists.toString();
+  }
 
   public synchronized boolean requestSplit(final HRegion r) {
     // don't split regions that are blocking
Index: F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java
===================================================================
--- F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java	(revision 7615)
+++ F:/Grace/work/mainworkingstuff/Cloud/Hadoop/hbase-0.94.0/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java	(revision 7634)
@@ -3675,6 +3675,10 @@
     HServerLoad hsl = buildServerLoad();
     return hsl == null? null: hsl.getCoprocessors();
   }
+  
+  public boolean isShowQueueDump(){
+    return this.conf.getBoolean("hbase.regionserver.servlet.show.queuedump", true);
+  }
 
   /**
    * Register bean with platform management server
