diff --git hbase-server/src/test/java/org/apache/hadoop/hbase/util/TestMiniClusterLoadEncoded.java hbase-server/src/test/java/org/apache/hadoop/hbase/util/TestMiniClusterLoadEncoded.java
index c0326aa..979b498 100644
--- hbase-server/src/test/java/org/apache/hadoop/hbase/util/TestMiniClusterLoadEncoded.java
+++ hbase-server/src/test/java/org/apache/hadoop/hbase/util/TestMiniClusterLoadEncoded.java
@@ -49,13 +49,4 @@ public class TestMiniClusterLoadEncoded extends TestMiniClusterLoadParallel {
   public TestMiniClusterLoadEncoded(DataBlockEncoding encoding) {
     super(USE_MULTI_PUT, encoding);
   }
-
-  /**
-   * Use a smaller number of keys in in this test.
-   */
-  @Override
-  protected int numKeys() {
-    return 3000;
-  }
-
 }
diff --git hbase-server/src/test/java/org/apache/hadoop/hbase/util/TestMiniClusterLoadSequential.java hbase-server/src/test/java/org/apache/hadoop/hbase/util/TestMiniClusterLoadSequential.java
index a40002a..92fbd79 100644
--- hbase-server/src/test/java/org/apache/hadoop/hbase/util/TestMiniClusterLoadSequential.java
+++ hbase-server/src/test/java/org/apache/hadoop/hbase/util/TestMiniClusterLoadSequential.java
@@ -167,7 +167,7 @@ public class TestMiniClusterLoadSequential {
   }
 
   protected int numKeys() {
-    return 10000;
+    return 1000;
   }
 
   protected HColumnDescriptor getColumnDesc(HBaseAdmin admin)
