Index: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFile.java
===================================================================
--- hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFile.java	(revision 1551818)
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFile.java	(working copy)
@@ -152,10 +152,12 @@
   /** Minimum supported HFile format version */
   public static final int MIN_FORMAT_VERSION = 2;
 
-  /** Maximum supported HFile format version
-   */
+  /** Maximum supported HFile format version */
   public static final int MAX_FORMAT_VERSION = 3;
 
+  /** Minimum HFile format version with support for cell tags */
+  public static final int MIN_FORMAT_VERSION_WITH_TAGS = 3;
+
   /** Default compression name: none. */
   public final static String DEFAULT_COMPRESSION =
     DEFAULT_COMPRESSION_ALGORITHM.getName();
Index: hbase-server/src/main/java/org/apache/hadoop/hbase/security/visibility/VisibilityController.java
===================================================================
--- hbase-server/src/main/java/org/apache/hadoop/hbase/security/visibility/VisibilityController.java	(revision 1551818)
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/security/visibility/VisibilityController.java	(working copy)
@@ -72,6 +72,7 @@
 import org.apache.hadoop.hbase.exceptions.DeserializationException;
 import org.apache.hadoop.hbase.filter.Filter;
 import org.apache.hadoop.hbase.filter.FilterList;
+import org.apache.hadoop.hbase.io.hfile.HFile;
 import org.apache.hadoop.hbase.io.util.StreamUtils;
 import org.apache.hadoop.hbase.ipc.RequestContext;
 import org.apache.hadoop.hbase.master.MasterServices;
@@ -160,6 +161,14 @@
   @Override
   public void start(CoprocessorEnvironment env) throws IOException {
     this.conf = env.getConfiguration();
+    int version = HFile.getFormatVersion(this.conf);
+    if (version < HFile.MIN_FORMAT_VERSION_WITH_TAGS) {
+      LOG.fatal("Given HFile version is " + version
+          + ". Visibility labels feature needs a minimum HFile version of "
+          + HFile.MIN_FORMAT_VERSION_WITH_TAGS);
+      throw new RuntimeException("Visibility labels feature needs a minimum HFile version of "
+          + HFile.MIN_FORMAT_VERSION_WITH_TAGS);
+    }
     ZooKeeperWatcher zk = null;
     if (env instanceof MasterCoprocessorEnvironment) {
       // if running on HMaster
