Index: hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/AccessController.java
===================================================================
--- hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/AccessController.java	(revision 1552715)
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/AccessController.java	(working copy)
@@ -155,13 +155,6 @@
 
   void initialize(RegionCoprocessorEnvironment e) throws IOException {
     final HRegion region = e.getRegion();
-    canPersistCellACLs = HFile.getFormatVersion(e.getConfiguration()) >=
-      HFile.MIN_FORMAT_VERSION_WITH_TAGS;
-    if (!canPersistCellACLs) {
-      LOG.info("A minimum HFile version of " + HFile.MIN_FORMAT_VERSION_WITH_TAGS
-        + " is required to persist cell ACLs. Consider setting " + HFile.FORMAT_VERSION_KEY
-        + " accordingly.");
-    }
     Map<byte[], ListMultimap<String,TablePermission>> tables =
         AccessControlLists.loadAll(region);
     // For each table, write out the table's permissions to the respective
@@ -633,7 +626,13 @@
   /* ---- MasterObserver implementation ---- */
 
   public void start(CoprocessorEnvironment env) throws IOException {
-
+    canPersistCellACLs = HFile.getFormatVersion(env.getConfiguration()) >=
+      HFile.MIN_FORMAT_VERSION_WITH_TAGS;
+    if (!canPersistCellACLs) {
+      LOG.info("A minimum HFile version of " + HFile.MIN_FORMAT_VERSION_WITH_TAGS
+          + " is required to persist cell ACLs. Consider setting " + HFile.FORMAT_VERSION_KEY
+          + " accordingly.");
+    }
     ZooKeeperWatcher zk = null;
     if (env instanceof MasterCoprocessorEnvironment) {
       // if running on HMaster
