From 2d1d89d121d87acd35f57ef43fce1dd2de8de998 Mon Sep 17 00:00:00 2001
From: Andrew Purtell <apurtell@apache.org>
Date: Wed, 16 Jul 2014 15:01:56 -0700
Subject: [PATCH] HBASE-2217 VM OPTS for shell only

---
 bin/hbase | 1 +
 1 file changed, 1 insertion(+)

diff --git a/bin/hbase b/bin/hbase
index 721ab5a..c85a916 100755
--- a/bin/hbase
+++ b/bin/hbase
@@ -275,6 +275,7 @@ if [ "$COMMAND" = "shell" ] ; then
   else
     HBASE_OPTS="$HBASE_OPTS -Dhbase.ruby.sources=$HBASE_HOME/hbase-shell/src/main/ruby"
   fi
+  HBASE_OPTS="$HBASE_OPTS $HBASE_SHELL_OPTS"
   CLASS="org.jruby.Main -X+O ${JRUBY_OPTS} ${HBASE_HOME}/bin/hirb.rb"
 elif [ "$COMMAND" = "hbck" ] ; then
   CLASS='org.apache.hadoop.hbase.util.HBaseFsck'
-- 
1.9.1

