diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/fs/HFileSystem.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/fs/HFileSystem.java
index fe8b3b2..904b8a4 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/fs/HFileSystem.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/fs/HFileSystem.java
@@ -257,7 +257,7 @@ public class HFileSystem extends FilterFileSystem {
     Class<?> clazz = conf.getClass("fs." + uri.getScheme() + ".impl", null);
     if (clazz != null) {
       // This will be true for Hadoop 1.0, or 0.20.
-      fs = (FileSystem)ReflectionUtils.newInstance(clazz, conf);
+      fs = (FileSystem) org.apache.hadoop.util.ReflectionUtils.newInstance(clazz, conf);
       fs.initialize(uri, conf);
     } else {
       // For Hadoop 2.0, we have to go through FileSystem for the filesystem
