Index: src/java/org/apache/hadoop/hbase/master/HMaster.java
===================================================================
--- src/java/org/apache/hadoop/hbase/master/HMaster.java	(revision 629189)
+++ src/java/org/apache/hadoop/hbase/master/HMaster.java	(working copy)
@@ -273,8 +273,11 @@
         fs.mkdirs(rootdir);
         FSUtils.setVersion(fs, rootdir);
       } else if (!FSUtils.checkVersion(fs, rootdir)) {
-        throw new IOException("File system needs upgrade. Run " +
-          "the '${HBASE_HOME}/bin/hbase migrate' script");
+        // Output on stdout so user sees it in terminal.
+        String message = "File system needs upgrade. Run " +
+          "the '${HBASE_HOME}/bin/hbase migrate' script.";
+        System.out.println("WARNING! " + message + " Master shutting down...");
+        throw new IOException(message);
       }
 
       if (!fs.exists(rootRegionDir)) {
