Index: src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java
===================================================================
--- src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java	(revision 1238835)
+++ src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java	(working copy)
@@ -2430,9 +2430,11 @@
     }
 
     public void leaseExpired() {
-      LOG.info("Scanner " + this.scannerName + " lease expired");
       RegionScanner s = scanners.remove(this.scannerName);
       if (s != null) {
+        LOG.info("Scanner " + this.scannerName + " lease expired on region "
+            + s.getRegionInfo().getRegionNameAsString() + ", table "
+            + s.getRegionInfo().getTableNameAsString());
         try {
           HRegion region = getRegion(s.getRegionInfo().getRegionName());
           if (region != null && region.getCoprocessorHost() != null) {
@@ -2447,6 +2449,8 @@
           LOG.error("Closing scanner for "
               + s.getRegionInfo().getRegionNameAsString(), e);
         }
+      } else{
+        LOG.info("Scanner " + this.scannerName + " lease expired");
       }
     }
   }
