diff --git a/hbase-common/src/main/java/org/apache/hadoop/hbase/ArrayBackedTag.java b/hbase-common/src/main/java/org/apache/hadoop/hbase/ArrayBackedTag.java
index 12f7a15..ee3cdf3 100644
--- a/hbase-common/src/main/java/org/apache/hadoop/hbase/ArrayBackedTag.java
+++ b/hbase-common/src/main/java/org/apache/hadoop/hbase/ArrayBackedTag.java
@@ -28,7 +28,7 @@ import org.apache.hadoop.hbase.util.Bytes;
 /**
  * This is a {@link Tag} implementation in which value is backed by an on heap byte array.
  */
-@InterfaceAudience.Private
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.COPROC)
 @InterfaceStability.Evolving
 public class ArrayBackedTag implements Tag {
   private final byte type;// TODO  extra type state needed?
diff --git a/hbase-common/src/main/java/org/apache/hadoop/hbase/ByteBufferTag.java b/hbase-common/src/main/java/org/apache/hadoop/hbase/ByteBufferTag.java
index 3b27fc0..a8a370d 100644
--- a/hbase-common/src/main/java/org/apache/hadoop/hbase/ByteBufferTag.java
+++ b/hbase-common/src/main/java/org/apache/hadoop/hbase/ByteBufferTag.java
@@ -29,7 +29,7 @@ import org.apache.hadoop.hbase.util.ByteBufferUtils;
  * This is a {@link Tag} implementation in which value is backed by
  * {@link java.nio.ByteBuffer}
  */
-@InterfaceAudience.Private
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.COPROC)
 @InterfaceStability.Evolving
 public class ByteBufferTag implements Tag {
 
diff --git a/hbase-common/src/main/java/org/apache/hadoop/hbase/Tag.java b/hbase-common/src/main/java/org/apache/hadoop/hbase/Tag.java
index 8a25898..55efe60 100644
--- a/hbase-common/src/main/java/org/apache/hadoop/hbase/Tag.java
+++ b/hbase-common/src/main/java/org/apache/hadoop/hbase/Tag.java
@@ -33,7 +33,7 @@ import org.apache.hadoop.hbase.util.Bytes;
  * <p>
  * See {@link TagType} for reserved tag types.
  */
-@InterfaceAudience.Private
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.COPROC)
 @InterfaceStability.Evolving
 public interface Tag {
 
diff --git a/hbase-common/src/main/java/org/apache/hadoop/hbase/TagType.java b/hbase-common/src/main/java/org/apache/hadoop/hbase/TagType.java
index 71a2fbb..cb334f9 100644
--- a/hbase-common/src/main/java/org/apache/hadoop/hbase/TagType.java
+++ b/hbase-common/src/main/java/org/apache/hadoop/hbase/TagType.java
@@ -20,7 +20,7 @@ package org.apache.hadoop.hbase;
 import org.apache.yetus.audience.InterfaceAudience;
 import org.apache.yetus.audience.InterfaceStability;
 
-@InterfaceAudience.Private
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.COPROC)
 @InterfaceStability.Evolving
 public final class TagType {
   // Please declare new Tag Types here to avoid step on pre-existing tag types.
diff --git a/hbase-common/src/main/java/org/apache/hadoop/hbase/TagUtil.java b/hbase-common/src/main/java/org/apache/hadoop/hbase/TagUtil.java
index 0f010a0..598ea00 100644
--- a/hbase-common/src/main/java/org/apache/hadoop/hbase/TagUtil.java
+++ b/hbase-common/src/main/java/org/apache/hadoop/hbase/TagUtil.java
@@ -31,7 +31,8 @@ import org.apache.hadoop.hbase.util.ByteBufferUtils;
 import org.apache.hadoop.hbase.util.Bytes;
 import org.apache.hadoop.hbase.util.Pair;
 
-@InterfaceAudience.Private
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.COPROC)
+// TODO : Add more Util methods here like TagItertor, copyTagsTo from CellUtil
 public final class TagUtil {
 
   // If you would like to check the length of tags, please call {@link TagUtil#checkForTagsLength()}.
