diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionServerMetrics.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionServerMetrics.java
index fa634d1..1ba33ed 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionServerMetrics.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionServerMetrics.java
@@ -459,9 +459,9 @@ public void testMobMetrics() throws IOException, InterruptedException {
     ((HRegion)region).initialize();
     region.compact(true);
     metricsRegionServer.getRegionServerWrapper().forceRecompute();
-    metricsHelper.assertCounter("mobCompactedFromMobCellsCount", numHfiles,
+    metricsHelper.assertCounter("cellsCountCompactedFromMob", numHfiles,
         serverSource);
-    metricsHelper.assertCounter("mobCompactedIntoMobCellsCount", 0, serverSource);
+    metricsHelper.assertCounter("cellsCountCompactedToMob", 0, serverSource);
     scanner = t.getScanner(scan);
     scanner.next(100);
     numScanNext++;  // this is an ugly construct
@@ -480,8 +480,8 @@ public void testMobMetrics() throws IOException, InterruptedException {
     region.compact(true);
     metricsRegionServer.getRegionServerWrapper().forceRecompute();
     // metrics are reset by the region initialization
-    metricsHelper.assertCounter("mobCompactedFromMobCellsCount", 0, serverSource);
-    metricsHelper.assertCounter("mobCompactedIntoMobCellsCount", 2 * numHfiles - 1,
+    metricsHelper.assertCounter("cellsCountCompactedFromMob", 0, serverSource);
+    metricsHelper.assertCounter("cellsCountCompactedToMob", 2 * numHfiles - 1,
         serverSource);
     t.close();
     admin.close();
