From 9b8d7e0aefe29fea474e06ab21f9288245ac3bf5 Mon Sep 17 00:00:00 2001
From: Michael Stack <stack@apache.org>
Date: Mon, 29 Jan 2018 22:14:47 -0800
Subject: [PATCH] HBASE-19891 Up nightly test run timeout from 6 hours to 8

---
 dev-support/Jenkinsfile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dev-support/Jenkinsfile b/dev-support/Jenkinsfile
index 14092d5da6..c7adfa5040 100644
--- a/dev-support/Jenkinsfile
+++ b/dev-support/Jenkinsfile
@@ -25,7 +25,7 @@ pipeline {
   }
   options {
     buildDiscarder(logRotator(numToKeepStr: '30'))
-    timeout (time: 6, unit: 'HOURS')
+    timeout (time: 8, unit: 'HOURS')
     timestamps()
     skipDefaultCheckout()
   }
-- 
2.11.0 (Apple Git-81)
