diff --git hbase-common/src/main/java/org/apache/hadoop/hbase/Cell.java hbase-common/src/main/java/org/apache/hadoop/hbase/Cell.java
index a82458b..dcbee8a 100644
--- hbase-common/src/main/java/org/apache/hadoop/hbase/Cell.java
+++ hbase-common/src/main/java/org/apache/hadoop/hbase/Cell.java
@@ -58,7 +58,7 @@ import org.apache.hadoop.classification.InterfaceStability;
  * byte[]'s.
  * <p/>
  */
-@InterfaceAudience.Private
+@InterfaceAudience.Public
 @InterfaceStability.Evolving
 public interface Cell {
 
