diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CombinedBlockCache.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CombinedBlockCache.java
index c89073d..cc53e60 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CombinedBlockCache.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CombinedBlockCache.java
@@ -36,7 +36,7 @@ import org.apache.hadoop.hbase.io.hfile.bucket.BucketCache;
  * 
  */
 @InterfaceAudience.Private
-public class CombinedBlockCache implements BlockCache, HeapSize {
+public class CombinedBlockCache implements ResizableBlockCache, HeapSize {
   private final LruBlockCache lruCache;
   private final BucketCache bucketCache;
   private final CombinedCacheStats combinedCacheStats;
@@ -207,4 +207,9 @@ public class CombinedBlockCache implements BlockCache, HeapSize {
   public BlockCache[] getBlockCaches() {
     return new BlockCache [] {this.lruCache, this.bucketCache};
   }
+
+  @Override
+  public void setMaxSize(long size) {
+    this.lruCache.setMaxSize(size);
+  }
 }
\ No newline at end of file
