Index: src/java/org/apache/hadoop/hbase/KeyValue.java
===================================================================
--- src/java/org/apache/hadoop/hbase/KeyValue.java	(revision 880882)
+++ src/java/org/apache/hadoop/hbase/KeyValue.java	(working copy)
@@ -200,6 +200,7 @@
   private byte [] bytes = null;
   private int offset = 0;
   private int length = 0;
+  private int keylength;
 
   /** Writable Constructor -- DO NOT USE */
   public KeyValue() {}
@@ -235,6 +236,7 @@
     this.bytes = bytes;
     this.offset = offset;
     this.length = length;
+    this.keylength = calculateKeyLength();
   }
 
   /** Constructors that build a new backing byte array from fields */
@@ -379,6 +381,7 @@
         timestamp, type, value, voffset, vlength);
     this.length = bytes.length;
     this.offset = 0;
+    this.keylength = calculateKeyLength();
   }
 
   /**
@@ -651,6 +654,10 @@
    * @return Length of key portion.
    */
   public int getKeyLength() {
+    return this.keylength;
+  }
+
+  private int calculateKeyLength() {
     return Bytes.toInt(this.bytes, this.offset);
   }
 
@@ -1783,6 +1790,7 @@
     this.offset = 0;
     this.bytes = new byte[this.length];
     in.readFully(this.bytes, 0, this.length);
+    this.keylength = calculateKeyLength();
   }
   
   public void write(final DataOutput out) throws IOException {
