diff --git a/hbase-common/src/main/resources/hbase-default.xml b/hbase-common/src/main/resources/hbase-default.xml
index 9931b2d..f7288c8 100644
--- a/hbase-common/src/main/resources/hbase-default.xml
+++ b/hbase-common/src/main/resources/hbase-default.xml
@@ -1572,9 +1572,9 @@ possible configurations would overwhelm and obscure the important.
   </property>
   <property>
     <name>hbase.master.mob.ttl.cleaner.period</name>
-    <value>86400000</value>
+    <value>86400</value>
     <description>
-      The period that ExpiredMobFileCleanerChore runs. The unit is millisecond.
+      The period that ExpiredMobFileCleanerChore runs. The unit is second.
       The default value is one day.
     </description>
   </property>
@@ -1609,9 +1609,9 @@ possible configurations would overwhelm and obscure the important.
   </property>
   <property>
     <name>hbase.mob.file.compaction.chore.period</name>
-    <value>604800000</value>
+    <value>604800</value>
     <description>
-      The period that MobFileCompactionChore runs. The unit is millisecond.
+      The period that MobFileCompactionChore runs. The unit is second.
       The default value is one week.
     </description>
   </property>
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/ExpiredMobFileCleanerChore.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/ExpiredMobFileCleanerChore.java
index a9e9773..7b06462 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/ExpiredMobFileCleanerChore.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/ExpiredMobFileCleanerChore.java
@@ -20,6 +20,7 @@
 
 import java.io.IOException;
 import java.util.Map;
+import java.util.concurrent.TimeUnit;
 
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
@@ -47,9 +48,10 @@
   private ExpiredMobFileCleaner cleaner;
 
   public ExpiredMobFileCleanerChore(HMaster master) {
-    super(master.getServerName() + "-ExpiredMobFileCleanerChore", master,
-        master.getConfiguration().getInt(MobConstants.MOB_CLEANER_PERIOD,
-                MobConstants.DEFAULT_MOB_CLEANER_PERIOD));
+    super(master.getServerName() + "-ExpiredMobFileCleanerChore", master, master.getConfiguration()
+      .getInt(MobConstants.MOB_CLEANER_PERIOD, MobConstants.DEFAULT_MOB_CLEANER_PERIOD), master
+      .getConfiguration().getInt(MobConstants.MOB_CLEANER_PERIOD,
+        MobConstants.DEFAULT_MOB_CLEANER_PERIOD), TimeUnit.SECONDS);
     this.master = master;
     this.tableLockManager = master.getTableLockManager();
     cleaner = new ExpiredMobFileCleaner();
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCompactionChore.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCompactionChore.java
index 3a12725..13c52f0 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCompactionChore.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCompactionChore.java
@@ -20,6 +20,7 @@
 
 import java.util.Map;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
 
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
@@ -44,9 +45,11 @@
   private ExecutorService pool;
 
   public MobFileCompactionChore(HMaster master) {
-    super(master.getServerName() + "-MobFileCompactChore", master,
-        master.getConfiguration().getInt(MobConstants.MOB_FILE_COMPACTION_CHORE_PERIOD,
-      MobConstants.DEFAULT_MOB_FILE_COMPACTION_CHORE_PERIOD));
+    super(master.getServerName() + "-MobFileCompactChore", master, master.getConfiguration()
+      .getInt(MobConstants.MOB_FILE_COMPACTION_CHORE_PERIOD,
+        MobConstants.DEFAULT_MOB_FILE_COMPACTION_CHORE_PERIOD), master.getConfiguration().getInt(
+      MobConstants.MOB_FILE_COMPACTION_CHORE_PERIOD,
+      MobConstants.DEFAULT_MOB_FILE_COMPACTION_CHORE_PERIOD), TimeUnit.SECONDS);
     this.master = master;
     this.tableLockManager = master.getTableLockManager();
     this.pool = MobUtils.createMobFileCompactorThreadPool(master.getConfiguration());
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobConstants.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobConstants.java
index 13c6b3c..464a0e7 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobConstants.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobConstants.java
@@ -44,7 +44,7 @@
   public static final byte[] MOB_REGION_NAME_BYTES = Bytes.toBytes(MOB_REGION_NAME);
 
   public static final String MOB_CLEANER_PERIOD = "hbase.master.mob.ttl.cleaner.period";
-  public static final int DEFAULT_MOB_CLEANER_PERIOD = 24 * 60 * 60 * 1000; // one day
+  public static final int DEFAULT_MOB_CLEANER_PERIOD = 24 * 60 * 60; // one day
 
   public static final String MOB_SWEEP_TOOL_COMPACTION_START_DATE =
       "hbase.mob.sweep.tool.compaction.start.date";
@@ -106,7 +106,7 @@
   public static final String MOB_FILE_COMPACTION_CHORE_PERIOD =
     "hbase.mob.file.compaction.chore.period";
   public static final int DEFAULT_MOB_FILE_COMPACTION_CHORE_PERIOD =
-    24 * 60 * 60 * 1000 * 7; // a week
+    24 * 60 * 60 * 7; // a week
   public static final String MOB_FILE_COMPACTOR_CLASS_KEY = "hbase.mob.file.compactor.class";
   /**
    * The max number of threads used in MobFileCompactor.
