Index: src/main/java/org/apache/hadoop/hbase/regionserver/wal/SequenceFileLogWriter.java
===================================================================
--- src/main/java/org/apache/hadoop/hbase/regionserver/wal/SequenceFileLogWriter.java	(revision 1325551)
+++ src/main/java/org/apache/hadoop/hbase/regionserver/wal/SequenceFileLogWriter.java	(working copy)
@@ -130,7 +130,7 @@
   public void init(FileSystem fs, Path path, Configuration conf)
   throws IOException {
     // Should we do our custom WAL compression?
-    boolean compress = conf.getBoolean(HConstants.ENABLE_WAL_COMPRESSION, false);
+    boolean compress = conf.getBoolean(HConstants.ENABLE_WAL_COMPRESSION, true);
     if (compress) {
       try {
         if (this.compressionContext == null) {
