From 364650bad519e3966dd8033464413c82e75eee06 Mon Sep 17 00:00:00 2001
From: Michael Stack <stack@apache.org>
Date: Mon, 29 Jan 2018 21:33:16 -0800
Subject: [PATCH] HBASE-19868 TestCoprocessorWhitelistMasterObserver is flakey

---
 .../security/access/TestCoprocessorWhitelistMasterObserver.java   | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/security/access/TestCoprocessorWhitelistMasterObserver.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/security/access/TestCoprocessorWhitelistMasterObserver.java
index 1686ba37aa..c161b9e38e 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/security/access/TestCoprocessorWhitelistMasterObserver.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/security/access/TestCoprocessorWhitelistMasterObserver.java
@@ -103,7 +103,7 @@ public class TestCoprocessorWhitelistMasterObserver extends SecureTestUtil {
         CoprocessorWhitelistMasterObserver.CP_COPROCESSOR_WHITELIST_PATHS_KEY,
         whitelistedPaths);
     // set retries low to raise exception quickly
-    conf.setInt("hbase.client.retries.number", 1);
+    conf.setInt("hbase.client.retries.number", 5);
     UTIL.startMiniCluster();
     UTIL.createTable(TEST_TABLE, new byte[][] { TEST_FAMILY });
     UTIL.waitUntilAllRegionsAssigned(TEST_TABLE);
@@ -137,7 +137,7 @@ public class TestCoprocessorWhitelistMasterObserver extends SecureTestUtil {
   private static void negativeTestCase(String[] whitelistedPaths,
       String coprocessorPath) throws Exception {
     Configuration conf = UTIL.getConfiguration();
-    conf.setInt("hbase.client.retries.number", 1);
+    conf.setInt("hbase.client.retries.number", 5);
     // load coprocessor under test
     conf.set(CoprocessorHost.MASTER_COPROCESSOR_CONF_KEY,
         CoprocessorWhitelistMasterObserver.class.getName());
@@ -269,7 +269,7 @@ public class TestCoprocessorWhitelistMasterObserver extends SecureTestUtil {
     conf.setStrings(CoprocessorWhitelistMasterObserver.CP_COPROCESSOR_WHITELIST_PATHS_KEY,
         new String[]{});
     // set retries low to raise exception quickly
-    conf.setInt("hbase.client.retries.number", 1);
+    conf.setInt("hbase.client.retries.number", 5);
     UTIL.startMiniCluster();
     HTableDescriptor htd = new HTableDescriptor(TEST_TABLE);
     HColumnDescriptor hcd = new HColumnDescriptor(TEST_FAMILY);
@@ -314,7 +314,7 @@ public class TestCoprocessorWhitelistMasterObserver extends SecureTestUtil {
     conf.setStrings(CoprocessorWhitelistMasterObserver.CP_COPROCESSOR_WHITELIST_PATHS_KEY,
         new String[]{});
     // set retries low to raise exception quickly
-    conf.setInt("hbase.client.retries.number", 1);
+    conf.setInt("hbase.client.retries.number", 5);
     UTIL.startMiniCluster();
     HTableDescriptor htd = new HTableDescriptor(TEST_TABLE);
     HColumnDescriptor hcd = new HColumnDescriptor(TEST_FAMILY);
-- 
2.11.0 (Apple Git-81)
