diff --git a/src/java/org/apache/hadoop/hbase/io/HbaseObjectWritable.java b/src/java/org/apache/hadoop/hbase/io/HbaseObjectWritable.java
index fde0a19..2fdb3d4 100644
--- a/src/java/org/apache/hadoop/hbase/io/HbaseObjectWritable.java
+++ b/src/java/org/apache/hadoop/hbase/io/HbaseObjectWritable.java
@@ -104,6 +104,7 @@ public class HbaseObjectWritable implements Writable, Configurable {
     addToMap(byte [][].class, code++);
     
     // Hadoop types
+    addToMap(Text.class, code++);
     addToMap(Writable.class, code++);
     addToMap(Writable [].class, code++);
     addToMap(HbaseMapWritable.class, code++);
