diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobUtils.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobUtils.java
index 149eed4..c106e0b 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobUtils.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobUtils.java
@@ -251,7 +251,7 @@ public static KeyValue createMobRefKeyValue(KeyValue kv, byte[] fileName, Tag ta
     // find the original mob files by this table name. For details please see cloning
     // snapshot for mob files.
     existingTags.add(tableNameTag);
-    long valueLength = kv.getValueLength();
+    int valueLength = kv.getValueLength();
     byte[] refValue = Bytes.add(Bytes.toBytes(valueLength), fileName);
     KeyValue reference = new KeyValue(kv.getRowArray(), kv.getRowOffset(), kv.getRowLength(),
         kv.getFamilyArray(), kv.getFamilyOffset(), kv.getFamilyLength(), kv.getQualifierArray(),
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HMobStore.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HMobStore.java
index 6b4e450..08157f5 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HMobStore.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HMobStore.java
@@ -219,9 +219,9 @@ private void validateMobFile(Path path) throws IOException {
    */
   public Cell resolve(Cell reference, boolean cacheBlocks) throws IOException {
     Cell result = null;
-    if (reference.getValueLength() > Bytes.SIZEOF_LONG) {
+    if (reference.getValueLength() > Bytes.SIZEOF_INT) {
       String fileName = Bytes.toString(reference.getValueArray(), reference.getValueOffset()
-          + Bytes.SIZEOF_LONG, reference.getValueLength() - Bytes.SIZEOF_LONG);
+          + Bytes.SIZEOF_INT, reference.getValueLength() - Bytes.SIZEOF_INT);
       Path targetPath = new Path(mobFamilyPath, fileName);
       MobFile file = null;
       try {
diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestDeleteMobTable.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestDeleteMobTable.java
index ed4b703..f2b1a62 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestDeleteMobTable.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestDeleteMobTable.java
@@ -213,7 +213,7 @@ private String assertHasOneMobRow(HTable table, TableName tn, String familyName)
     Result r = rs.next();
     Assert.assertNotNull(r);
     byte[] value = r.getValue(FAMILY, QF);
-    String fileName = Bytes.toString(value, Bytes.SIZEOF_LONG, value.length - Bytes.SIZEOF_LONG);
+    String fileName = Bytes.toString(value, Bytes.SIZEOF_INT, value.length - Bytes.SIZEOF_INT);
     Path filePath = new Path(
         MobUtils.getMobFamilyPath(TEST_UTIL.getConfiguration(), tn, familyName), fileName);
     FileSystem fs = TEST_UTIL.getTestFileSystem();
diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestHMobStore.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestHMobStore.java
index c253611..a7ae8cc 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestHMobStore.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestHMobStore.java
@@ -169,9 +169,9 @@ private void init(Configuration conf, HColumnDescriptor hcd)
     mobWriter.append(key3);
     mobWriter.close();
 
-    long valueLength1 = key1.getValueLength();
-    long valueLength2 = key2.getValueLength();
-    long valueLength3 = key3.getValueLength();
+    int valueLength1 = key1.getValueLength();
+    int valueLength2 = key2.getValueLength();
+    int valueLength3 = key3.getValueLength();
 
     String targetPathName = MobUtils.formatDate(currentDate);
     byte[] referenceValue =
diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestMobStoreScanner.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestMobStoreScanner.java
index 2af97c3..69b9c8f 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestMobStoreScanner.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestMobStoreScanner.java
@@ -307,7 +307,10 @@ private static void assertIsMobReference(Cell cell, byte[] row, byte[] family,
     Assert.assertFalse(Bytes.toString(value).equals(
         Bytes.toString(CellUtil.cloneValue(cell))));
     byte[] referenceValue = CellUtil.cloneValue(cell);
-    String fileName = Bytes.toString(referenceValue, 8, referenceValue.length-8);
+    String fileName = Bytes.toString(referenceValue, Bytes.SIZEOF_INT,
+        referenceValue.length - Bytes.SIZEOF_INT);
+    int valLen = Bytes.toInt(referenceValue, 0, Bytes.SIZEOF_INT);
+    Assert.assertEquals(value.length, valLen);
     Path mobFamilyPath;
     mobFamilyPath = new Path(MobUtils.getMobRegionPath(TEST_UTIL.getConfiguration(),
         TableName.valueOf(TN)), hcd.getNameAsString());
