Index: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSink.java
===================================================================
--- hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSink.java	(revision 1388119)
+++ hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSink.java	(working copy)
@@ -101,9 +101,9 @@
    */
   private void decorateConf() {
     this.conf.setInt(HConstants.HBASE_CLIENT_RETRIES_NUMBER,
-        this.conf.getInt("replication.sink.client.retries.number", 1));
+        this.conf.getInt("replication.sink.client.retries.number", 4));
     this.conf.setInt(HConstants.HBASE_CLIENT_OPERATION_TIMEOUT,
-        this.conf.getInt("replication.sink.client.ops.timeout", 20));
+        this.conf.getInt("replication.sink.client.ops.timeout", 10000));
    }
 
   /**
