Index: src/main/java/org/apache/hadoop/hbase/regionserver/Store.java
===================================================================
--- src/main/java/org/apache/hadoop/hbase/regionserver/Store.java	(revision 1042756)
+++ src/main/java/org/apache/hadoop/hbase/regionserver/Store.java	(working copy)
@@ -197,7 +197,8 @@
     this.majorCompactionTime = getNextMajorCompactTime();
 
     this.maxFilesToCompact = conf.getInt("hbase.hstore.compaction.max", 10);
-    this.minCompactSize = this.region.memstoreFlushSize * 3 / 2; // +50% pad
+    this.minCompactSize = conf.getInt("hbase.hstore.compaction.minsize",
+        (int)(this.region.memstoreFlushSize * 3 / 2)); // +50% pad
     this.compactRatio = conf.getFloat("hbase.hstore.compaction.ratio", 1.2F);
 
     if (Store.closeCheckInterval == 0) {
