From 1b4bee7a5b1bc7ff8397b41331a94287bbc4d8f6 Mon Sep 17 00:00:00 2001
From: Sean Busbey <busbey@apache.org>
Date: Thu, 12 Apr 2018 16:31:24 -0500
Subject: [PATCH] HBASE-20379 shadedjars yetus plugin should add a footer link

---
 dev-support/hbase-personality.sh | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dev-support/hbase-personality.sh b/dev-support/hbase-personality.sh
index 3507a1d8de..219891387f 100755
--- a/dev-support/hbase-personality.sh
+++ b/dev-support/hbase-personality.sh
@@ -365,6 +365,7 @@ function shadedjars_rebuild
   count=$(${GREP} -c '\[ERROR\]' "${logfile}")
   if [[ ${count} -gt 0 ]]; then
     add_vote_table -1 shadedjars "${repostatus} has ${count} errors when building our shaded downstream artifacts."
+    add_footer_table shadedjars "@@BASE@@/${repostatus}-shadedjars.txt"
     return 1
   fi
 
-- 
2.16.1

