### Eclipse Workspace Patch 1.0
#P 0.94.0-ali-1.0
Index: src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon
===================================================================
--- src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon	(revision 210095)
+++ src/main/jamon/org/apache/hadoop/hbase/tmpl/master/MasterStatusTmpl.jamon	(working copy)
@@ -24,6 +24,7 @@
 ServerName rootLocation = null;
 ServerName metaLocation = null;
 List<ServerName> servers = null;
+List<GroupInfo> groups = null;
 Set<ServerName> deadServers = null;
 boolean showAppendWarning = false;
 String filter = "general";
@@ -43,6 +44,8 @@
 org.apache.hadoop.hbase.client.HConnectionManager;
 org.apache.hadoop.hbase.HTableDescriptor;
 org.apache.hadoop.hbase.HBaseConfiguration;
+org.apache.hadoop.hbase.group.GroupInfo;
+com.google.common.base.Joiner;
 </%import>
 <%if format.equals("json") %>
   <& ../common/TaskMonitorTmpl; filter = filter; format = "json" &>
@@ -126,6 +129,9 @@
 <%if (servers != null) %>
 <& regionServers &>
 </%if>
+<%if (groups != null) %>
+<& groupInfos &>
+</%if>
 <%if (deadServers != null) %>
 <& deadRegionServers &>
 </%if>
@@ -232,6 +238,26 @@
 </%if>
 </%def>
 
+<%def groupInfos>
+<h2>Server Group Informations</h2>
+<%if (groups != null && groups.size() > 0)%>
+<table>
+<tr><th>Group Name</th><th>Servers</th><th>Tables</th></tr>
+<%java>
+  for(GroupInfo group: groups){
+    String servers = Joiner.on(',').join(group.getServers());
+    String tables = Joiner.on(',').join(group.getTables());
+    String groupName = group.getName();
+</%java>    
+<tr><td><% groupName %></td><td><% servers %></td><td><% tables %></td></tr>
+<%java>
+  }
+</%java>
+<tr><th>Total:<% groups.size() %></th><td></td><td></td></tr>
+</table>
+</%if>
+</%def>
+
 <%def deadRegionServers>
 <h3>Dead Region Servers</h3>
 <%if (deadServers != null && deadServers.size() > 0)%>
Index: src/main/java/org/apache/hadoop/hbase/master/MasterStatusServlet.java
===================================================================
--- src/main/java/org/apache/hadoop/hbase/master/MasterStatusServlet.java	(revision 210095)
+++ src/main/java/org/apache/hadoop/hbase/master/MasterStatusServlet.java	(working copy)
@@ -20,7 +20,6 @@
 package org.apache.hadoop.hbase.master;
 
 import java.io.IOException;
-
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
@@ -34,8 +33,9 @@
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.hbase.ServerName;
 import org.apache.hadoop.hbase.client.HBaseAdmin;
+import org.apache.hadoop.hbase.group.GroupBasedLoadBalancer;
+import org.apache.hadoop.hbase.tmpl.master.MasterStatusTmpl;
 import org.apache.hadoop.hbase.util.FSUtils;
-import org.apache.hadoop.hbase.tmpl.master.MasterStatusTmpl;
 
 /**
  * The servlet responsible for rendering the index page of the
@@ -70,6 +70,10 @@
       .setMetaLocation(metaLocation)
       .setServers(servers)
       .setDeadServers(deadServers);
+    if (master.getLoadBalancer() instanceof GroupBasedLoadBalancer) {
+      tmpl.setGroups(((GroupBasedLoadBalancer) master.getLoadBalancer())
+          .getGroupInfoManager().listGroups());
+    }
     if (request.getParameter("filter") != null)
       tmpl.setFilter(request.getParameter("filter"));
     if (request.getParameter("format") != null)
